packages/light-client/src/block.ts

```diff
@@ -0,0 +1,165 @@
+import { computeBlockHash, combineHash } from './utils';
+import {
+    SCHEMA,
+    BorshApprovalInner,
+    BorshValidatorStakeView,
+    BorshValidatorStakeViewV1,
+    BorshValidatorStakeViewWrapper,
+} from './borsh';
+import bs58 from 'bs58';
+import { sha256 } from 'js-sha256';
+import {
+    LightClientBlockLiteView,
+    NextLightClientBlockResponse,
+    ValidatorStakeView,
+} from '@near-js/types';
+import { PublicKey } from '@near-js/crypto';
+import BN from 'bn.js';
+import { serialize } from 'borsh';
+
+export interface ValidateLightClientBlockParams {
+    lastKnownBlock: LightClientBlockLiteView;
+    currentBlockProducers: ValidatorStakeView[];
+    newBlock: NextLightClientBlockResponse;
+}
+
+/**
+ * Validates a light client block response from the RPC against the last known block and block
+ * producer set.
+ *
+ * @param lastKnownBlock The last light client block retrieved. This must be the block at the epoch before newBlock.
+ * @param currentBlockProducers The block producer set for the epoch of the last known block.
+ * @param newBlock The new block to validate.
+ */
+export function validateLightClientBlock({
+    lastKnownBlock,
+    currentBlockProducers,
+    newBlock,
+}: ValidateLightClientBlockParams) {
+    // Numbers for each step references the spec:
+    // https://github.com/near/NEPs/blob/c7d72138117ed0ab86629a27d1f84e9cce80848f/specs/ChainSpec/LightClient.md
+    // (1) Verify that the block height is greater than the last known block.
+    if (newBlock.inner_lite.height <= lastKnownBlock.inner_lite.height) {
+        throw new Error(
+            'New block must be at least the height of the last known block'
+        );
+    }
+
+    // (2) Verify that the new block is in the same epoch or in the next epoch known to the last
+    // known block.
+    if (
+        newBlock.inner_lite.epoch_id !== lastKnownBlock.inner_lite.epoch_id &&
+        newBlock.inner_lite.epoch_id !== lastKnownBlock.inner_lite.next_epoch_id
+    ) {
+        throw new Error(
+            'New block must either be in the same epoch or the next epoch from the last known block'
+        );
+    }
+
+    const blockProducers: ValidatorStakeView[] = currentBlockProducers;
+    if (newBlock.approvals_after_next.length < blockProducers.length) {
```

Comment:
That implementation is wrong. The spec says, and in practice, there are blocks with more block producers than approvals because in the case of validator rotations, both the old and new bps must be included.

I can find and give a block where this is the case if needed, but I don't have an example off-hand.

Reply:
References, for posterity, to remove the trust aspect for others 😄 : https://github.com/near/NEPs/blob/master/specs/ChainSpec/LightClient.md#signature-verification. Also, line in rainbow bridge implementation, which is used in production https://github.com/aurora-is-near/rainbow-bridge/blob/33ca808b45cb5e9cf2e27f741b0f6e42d97c276b/contracts/eth/nearbridge/contracts/NearBridge.sol#L212 (python client just used for tests)

Also, anyone would be able to switch this to strict equal and see that it fails on recent blocks.